src/main/webapp/app/overview/course-learning-goals/course-learning-goals-details.component.ts

```diff
@@ -0,0 +1,123 @@
+
+    faPencilAlt = faPencilAlt;
+    getIcon = getIcon;
+    getIconTooltip = getIconTooltip;
+
+    constructor(
+        private alertService: AlertService,
+        private activatedRoute: ActivatedRoute,
+        private learningGoalService: LearningGoalService,
+        private lectureUnitService: LectureUnitService,
+    ) {}
+
+    ngOnInit(): void {
+        this.activatedRoute.params.subscribe((params) => {
+            this.learningGoalId = +params['learningGoalId'];
+            this.courseId = +params['courseId'];
+            if (this.learningGoalId && this.courseId) {
+                this.loadData();
+            }
+        });
+    }
+
+    private loadData() {
+        this.isLoading = true;
+        this.learningGoalService.findById(this.learningGoalId!, this.courseId!).subscribe({
+            next: (resp) => {
+                this.learningGoal = resp.body!;
+                if (this.learningGoal && this.learningGoal.exercises) {
+                    // Add exercises as lecture units for display
+                    (this.learningGoal.lectureUnits ?? []).push(
+                        ...this.learningGoal.exercises.map((exercise) => {
+                            const exerciseUnit = new ExerciseUnit();
+                            exerciseUnit.id = exercise.id;
+                            exerciseUnit.exercise = exercise;
+                            return exerciseUnit as LectureUnit;
+                        }),
+                    );
+                }
+                this.isLoading = false;
+            },
+            error: (errorResponse: HttpErrorResponse) => onError(this.alertService, errorResponse),
+        });
+    }
+
+    showFireworksIfMastered() {
+        if (this.mastery >= 100 && !this.showFireworks) {
+            setTimeout(() => (this.showFireworks = true), 1000);
+            setTimeout(() => (this.showFireworks = false), 6000);
+        }
+    }
+
+    getUserProgress(): LearningGoalProgress {
+        if (this.learningGoal.userProgress?.length) {
+            return this.learningGoal.userProgress.first()!;
+        }
+        return { progress: 0, confidence: 0 } as LearningGoalProgress;
+    }
+
+    get progress(): number {
+        // The percentage of completed lecture units and participated exercises
+        return Math.round(this.getUserProgress().progress ?? 0);
+    }
+
+    get confidence(): number {
+        // Confidence level (average score in exercises) in proportion to the threshold value (max. 100 %)
+        // Example: If the student’s latest confidence level equals 60 % and the mastery threshold is set to 80 %, the ring would be 75 % full.
+        return Math.min(Math.round(((this.getUserProgress().confidence ?? 0) / (this.learningGoal.masteryThreshold ?? 100)) * 100), 100);
+    }
+
+    get mastery(): number {
+        // Advancement towards mastery as a weighted function of progress and confidence
+        const weight = 2 / 3;
```

Comment:
This weight is used in multiple places, should we convert this to a global constant?

Reply:
I am thinking about moving this calculation to the server in a follow-up (because we recently discussed a similar issue with score calculation in the client w.r.t. the mobile app development).